### DISCUSSION TITLE
ESP NOW : wifi tips

### DISCUSSION THREAD

**Discussion by PatLub85** [micropython/micropython#11110] [General]
Hi,
here is a small program to help those who want to use ESP NOW, without tearing their hair like me. All the information can be found on this page https://micropython-glenn20.readthedocs.io/en/latest/library/espnow.html#espnow-and-wifi-operation, but they are scattered.
In summary, to use ESP32 with ESP8266, the ESP8266 must use the MAC AP address of the ESP32 to send data.
For my tests, I have an ESP32 or ESP8266 for the server, and an ESP 8266 for the client.
If the server is an ESP32, it is imperative to use this command: e.add_peerpeer,ifidx=network.AP_IF, which requires using the MAc address of the AP interface.
Also to be sure that the two ESPs communicate, you have to block the Wifi channel (1 for me)
The server
```
import network
import espnow
import os
from time import sleep

class Esp_now:
    def __init__(self):
        self.e = espnow.ESPNow()
        self.peer =b'\xb4\xe6-E\x1ba' # address of the client
        self.sta=network.WLAN(network.STA_IF)
        self.ap = network.WLAN(network.AP_IF)

    def initialisation(self):
        self.sta, self.ap = self.wifi_reset()
        self.ap.active(True)
        self.ap.config(channel=1)        
        self.e.active(True)
        try: #Sometimes I have an error OSError -5
            if os.uname()[0] == 'esp32':
                self.e.add_peer(self.peer,ifidx=network.AP_IF) #  forces the AP interface and its Mac address
            else:
                self.e.add_peer(self.peer)
        except OSError as error:
            print('error',error)
            machine.reset()

    def wifi_reset(self):   # Reset wifi to AP_IF off, STA_IF on and disconnected
        self.sta = network.WLAN(network.STA_IF); self.sta.active(False)
        self.ap = network.WLAN(network.AP_IF); self.ap.active(False)
        self.sta.active(True)
        while not self.sta.active():
            sleep(0.1)
            self.sta.disconnect()   # For ESP8266
        while self.sta.isconnected():
            sleep(0.1)
        return self.sta, self.ap
    
    def esp(self,name):
        self.e.send(self.peer, str(name), True)
        
    def em_rec(self):
        if os.uname()[0] == 'esp8266':
            self.esp(os.uname()[0])
        else:
            self.esp(os.uname()[0])
        while True:
            print ("Waiting for client...")
            host, msg = self.e.recv()
            if msg == b'START':
                print (msg.decode())
                self.e.send(self.peer, "information to send", True)
            elif msg == b'STOP':   
                print(msg.decode())
                break

def run():
    test=Esp_now()
    test.initialisation()
    test.em_rec()

run()
```
The client
```
import network
import os
import espnow
from time import sleep

class Esp_now:
    def __init__(self):
        self.e = espnow.ESPNow()

    def initialisation(self):
        self.sta, self.ap = self.wifi_reset()
        self.ap.active(True)
        self.ap.config(channel=1)
        self.ap.active(False)
        self.e.active(True)

    def transmission(self,data):
        self.e.send(self.peer, data, True)

    def host(self): # if the transmitter is an 8266 ,the @mac is that of the STA interface
        # if it is an ESP32, it is that of the AP interface
        print('Wait message')
        host, msg = self.e.recv()
        self.peer=host
        self.add_peer(host)
            
    def add_peer(self,peer):
        try:
            self.e.add_peer(peer)
        except OSError as error:
            print('error',error)
            machine.reset()

    def wifi_reset(self):   # Reset wifi to AP_IF off, STA_IF on and disconnected
        self.sta = network.WLAN(network.STA_IF); self.sta.active(False)
        self.ap = network.WLAN(network.AP_IF); self.ap.active(False)
        self.sta.active(True)
        while not self.sta.active():
            sleep(0.1)
            self.sta.disconnect()   # For ESP8266
        while self.sta.isconnected():
            sleep(0.1)
        return self.sta, self.ap

    def em_rec(self,data):
        if data ==  ('START'):
            print(data)
            self.transmission(b'START')   
            try:
                host, msg = self.e.recv()
                print("Message received:",msg.decode())
            except OSError as error:
                print('error',error)
        elif data ==  ('STOP'):
            self.transmission(b'STOP')    
            print(data)

def run():
    test=Esp_now()
    test.initialisation()
    test.host()
    test.em_rec('START')
    test.em_rec('STOP')

run()
```
I hope these few lines will be useful to someone.

**Comment by davefes** [comment]
Always good to have code examples posted.  Have you seen Glenn20's ESPnow utils on Github?
https://github.com/glenn20/micropython-espnow-utils
His wifi.py in my experience is bullet-proof.  I use it in all my ESPNow links and intend to use it in "normal" ESP32 WiFi applications.

**Reply by PatLub85** [reply]
No, I hadn't seen those precious examples. 
I searched this site for ESP NOW questions and didn't find many questions.
It's a pity that the previous conversation did not continue : https://forum.micropython.org/viewtopic.php?f=18&t=9177&sid=15561229b8cbd8dad9ec67d8f9ea34f5&start=360
It would be interesting to hear feedback from more users.

**Comment by PatLub85** [comment]
Question for Glenn if he sees this question:
is there a description of the errors.
Sometimes I get an OSError -5 when I do an e.send()

**Comment by davefes** [comment]
Error 5 suggests input/output error.

e.send() should be a string or byte-string.  Think I had an error once when I tried to send an int or a float.

**Reply by PatLub85** [reply]
No, that's not the reason for me. All data is either binary or string.
Where do I find the description of the errors?

**Reply by davefes** [reply]
karfas suggested in this thread:
https://github.com/orgs/micropython/discussions/11068
when I was getting an error -113

https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/network/esp_now.html
near the bottom list a bunch of error codes but no numbers against them.

**Comment by PatLub85** [comment]
I found the specialized discussion for ESP NOW and I will continue on this one
https://github.com/glenn20/micropython-espnow-images/issues?page=1&q=is%3Aissue+is%3Aclosed

**Reply by glenn20** [reply]
I'm just copying in my answer to the issue above in case anyone finds this error here:

> Generally, the Espressif error codes raised are greater than 0x100 (see [esp_err.h](https://github.com/espressif/esp-idf/blob/master/components/esp_common/include/esp_err.h) and [esp_now.h](https://github.com/espressif/esp-idf/blob/master/components/esp_wifi/include/esp_now.h)). Small numbers like 5 and 6 indicate micropython system errors which follow the Linux errno conventions. 5 is EIO (Input/Output Error) and 6 is ENXIO (No such device or address).

> I can't see how these errors could be raised by the espnow module, but there are other places that can raise these errors codes, including the Wifi code (lwip), the filesystem code and bluetooth. I can't rule out these codes being returned by an Espressif ESPNOW API call, but they are certainly not expected. I can speculate that some underlying wifi code is raising the exception.
